text/0059-token-metadata-document.md

```diff
@@ -0,0 +1,504 @@
+in an appropriate way.
+
+This proposal differs from an earlier proposal [MCIP 40](https://github.com/mobilecoinfoundation/mcips/pull/0040), in that it is a stop-gap mechanism which
+does not post metadata to the blockchain.
+
+This proposal specifies how clients should dynamically discover token metadata, authenticate this data,
+and how they can go on to use this data.
+
+# Motivation
+[motivation]: #motivation
+
+When a client gets a `TxOut`, at the lowest level, both the `value` and `token_id` are `u64`
+numbers which can be decrypted from the blockchain records. However, it isn't appropriate to
+display the `token_id` number directly to the users, since they won't know what this means.
+
+Metadata about a `token_id` consists of all the data which helps clients figure out how they
+can appropriately represent value to the users. Tampering with this metadata can represent an attack
+to confuse or steal from the users, so there is considerable interest in ensuring that there
+is a robust way for clients to authenticate the metadata.
+
+If we do not support dynamic discovery of token metadata, then it must always be baked into clients,
+which is what we do today.
+
+Because some clients have relatively long upgrade times, this means that new
+tokens that are launched still cannot actually be used by users. This effectively hurts the
+time-to-market on any new token that is created.
+
+There are second order effects to this -- if tokens become "available" in clients at different
+points in time, then users can get bitten in a bad way. For instance, if a Moby user sent eUSD
+to a Signal user, and at the time, Moby supports eUSD but Signal does not, then the recipient
+cannot see the payment in their app, or send it anywhere, or send it back to the Moby user.
+This could lead to disputed payments, bad user experiences, etc.
+
+At the same time, putting token metadata on chain will require an enclave upgrade, while creating
+new tokens on the chain does not. So there is considerable interest in having an off-chain source
+of token metadata that can be validated by clients.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This proposal creates a source of truth for token metadata for tokens on the MobileCoin blockchain.
+
+Two documents are hosted on the MobileCoin foundation website:
+
+* `https://www.mobilecoin.com/token_metadata.json`: A JSON document containing token metadata
+* `https://www.mobilecoin.com/token_metadata.sig`: An ed25519 signature over the bytes of the .json document. This is binary data.
+
+The ed25519 signing key is be controlled by the MobileCoin foundation.
+We propose that the `MINTING_TRUST_ROOT` key is reused for this purpose.
+
+Minting trust root ed25519 public key hex bytes: `TODO`
+
+The `token_metadata.json` document has the following schema (example):
+
+```
+{
+    "schema_version": 1,
+    "signing_timestamp": "10020032320032",
```

Comment:
I think this should be changed to `updated_at` or something like that, since the signing happens after this gets changed.

Reply:
In the consensus nodes, the field is called `signed_at` but I feel like calling it a timestamp is better because it is actually a timestamp, `at` is less specific.

I tried to change the semantics so that this is the "approximate" time of signing. I don't think it's a big deal if the signing happens a bit later than the document is actually changed.